### PR TITLE
[Model] Add CSM-1B (sesame/csm-1b) streaming TTS support

### LINKED ISSUE #799: [Feature] Support Sesame CSM-1B streaming TTS in SGLang Omni
### Checklist
- [x] This is a feature request, not a question.
- [x] Written in English.

### Motivation

Add native support for **CSM-1B** ([sesame/csm-1b](https://huggingface.co/sesame/csm-1b), Apache-2.0): a Llama-3.2-1B backbone with a 31-step depth decoder over 32 Kyutai Mimi RVQ codebooks, decoded by the Mimi vocoder. It is a popular, natural-sounding 1B conversational model that today only has the reference repo and single-stream community forks, with no production batching or streaming.

I already have a working, additive implementation that follows the existing Higgs TTS pattern (OmniScheduler continuous batching, paged-KV Llama backbone, a Mimi vocoder stage): a new `models/csm_tts/` package, a one-line registry entry, an example config, and GPU-free unit tests. Validated on an RTX 3060: serves `/v1/audio/speech` streaming and non-streaming, single and concurrent, with lint and unit tests green. The one new piece is a Kyutai Mimi codec stage (first in the zoo, reusable by other Mimi-based models).

Is adding CSM-1B aligned with the roadmap? I am also raising a PR with the implementation.

### Related resources
- https://huggingface.co/sesame/csm-1b
- https://github.com/SesameAILabs/csm

### PR DESCRIPTION
**Pull request by Venkat2811**
Closes #799.

## Motivation

Adds end-to-end support for Sesame **CSM-1B** ([sesame/csm-1b](https://huggingface.co/sesame/csm-1b), Apache-2.0): a Llama-3.2-1B backbone that emits one frame per autoregressive step, a 31-step depth decoder over 32 Kyutai Mimi RVQ codebooks, and the Mimi vocoder. It is a popular, natural-sounding 1B conversational model that until now had only the reference repo and single-stream community forks, with no production batching or streaming.

## Design

Standard 3-stage Omni TTS pipeline, following the existing Higgs TTS layout (#428):

`preprocessing → tts_engine (OmniScheduler: paged KV, continuous batching, abort) → vocoder (streaming Mimi)`

- **Model** (`sglang_omni/models/csm_tts/`): the backbone composes SGLang's `LlamaForCausalLM` under paged/varlen attention; the 31-step depth decoder runs inside the scheduler-visible decode step. The codebooks head, frame embedding, sampler, Mimi codec wrapper, and weight remap are framework-free torch ported from HuggingFace Transformers `transformers/models/csm` and `transformers.MimiModel`.
- **Runner / scheduling**: `CsmTTSModelRunner` reuses the `ModelRunner` sync-decode collect and async-decode hooks (`post_decode_launch` / `post_decode_resolve`); the sampler-pool row discipline and CG-buffer population mirror the Higgs runner.
- **State / handoff**: `CsmTtsState` flows preprocessing → engine → vocoder via `StagePayload`.
- **Registration**: one line in `_register_omni_model` (`CsmForConditionalGeneration` → `csm_tts.model:CsmTTSModel`), auto-discovered via `EntryClass` + `architecture`.

The new codec piece is a Kyutai **Mimi** vocoder stage (RVQ at 12.5 Hz, 24 kHz mono), which appears to be the first Mimi backend in the repo and is reusable by other Mimi-based models. It slots into the existing TTS abstraction; it is not a new one.

## Scope

Minimal and additive: 26 files, registry is one line. Optimizations (CUDA-graph capture, depth-loop batching, async decode) are intentionally left as follow-ups to keep this first PR small. Async decode is implemented but off by default.

## Validation (RTX 3060, 12 GB)

- **Lint**: `pre-commit run --all-files` clean (autoflake + isort + black 24.10.0 + file hooks).
- **Unit tests**: 38 passed, 2 skipped (GPU-free). Covers HF-greedy depth bit-parity, the weight-remap census (the `lm_head`-is-codebook0 trap), async-decode sync-vs-async parity, top_p nucleus filtering, the EOS machine, and the out-of-bounds clamp fences.
- **End to end**: boots and serves `/v1/audio/speech` in streaming and non-streaming, single and concurrent (continuous batching confirmed to 8). A capped request returns a full clip (480000 PCM bytes, 125 frames at 24 kHz).

## Run

```bash
sgl-omni serve --model-path sesame/csm-1b --config examples/configs/csm_1b.yaml --port 8000
```

## Notes for reviewers

- `sesame/csm-1b` is gated; accept its license on HF before pulling weights.
- The example config is sized for a 12 GB card (eager, `mem_fraction_static 0.5`); flip the CUDA-graph and async-decode knobs once the optimization follow-ups land.

### CODE REVIEW

**Comment by Venkat2811** [comment]
@shuaills @FrankLeeeee could you please enable CI for this PR ?

**Comment by GaokaiZhang** [comment]
Reviewed the branch against `main` and ran the full EN seed-tts-eval on a single H100. The modeling core is careful, but it doesn't boot on Hopper as-is, and once patched to run, both WER and speed are well off where they need to be.

**Setup.** 1× H100 80GB, DP 1 (no router), shipped `examples/configs/csm_1b.yaml`, full EN set (1088), `--max-concurrency 16`, plain TTS (`--no-ref-audio`, the documented single-speaker path), WER via `whisper-large-v3`. It crashes on the first request out of the box (#1 below); the numbers are with the cuDNN SDPA backend disabled — the only change needed to boot. Weights from the ungated `eustlb/csm-1b` mirror (byte-identical 538-tensor HF-transformers export; `sesame/csm-1b` is gated to an authorized list).

| config | Split | WER / CER | Samples | Failed / skipped | Throughput | rtf_mean |
|---|---|---|---|---|---|---|
| as shipped (eager, frame cap 125) | en | **12.32%** | 1088 | 0 / 0 | 1.65 req/s | 2.29 |
| frame cap → 1024, else identical | en | **11.25%** | 1088 | 0 / 0 | 0.77 req/s | 3.40 |

Both are too high to be useful yet — WER ~12% (vs ~1–2% for the other TTS families on this eval) and rtf ~2.3 (>2× slower than real time, eager / no CUDA graph). The WER is a tail, not uniform: median **0%**, but **2.5% on the 898 clips that terminate vs 56% on the 190 (17%) that run to the 10 s frame cap** (95% of the >50% tail). Those are *short* sentences run long, so it's **non-termination, not truncation** — ~1 in 6 plain-TTS requests never emits EOS at the shipped defaults (temp 0.9 / top_k 50, no ref audio, no rep penalty). Cap → 1024 confirms it: the same ~17.6% still run away (now to 20–82 s), WER barely moves, throughput halves.

**What needs fixing, in order:**

1. **H100 crash (blocker).** The depth-decoder attention's `F.scaled_dot_product_attention` (`modeling.py:234`) dispatches to PyTorch's cuDNN SDPA backend on Hopper and aborts (`mha_graph.execute(...).is_good() == false`); the 3060 it was validated on picks a different backend. Fix: force a non-cuDNN backend — wrap that call in `with torch.nn.attention.sdpa_kernel([SDPBackend.FLASH_ATTENTION, SDPBackend.EFFICIENT_ATTENTION, SDPBackend.MATH]):`, or call `torch.backends.cuda.enable_cudnn_sdp(False)` once at model init. Every machine here is H100.
2. **Plain-TTS non-termination — the WER killer** (12% vs ~2.5% on clips that stop). Try a repetition penalty / lower default temperature / reference-audio context; at minimum warn when a request hits the frame cap and document the limit. Raising the cap is not the fix.
3. **rtf 2.3 — perf gate.** The deferred CUDA-graph / async-decode work is needed before this is servable. Note the toggle is currently oversold: `stages.py:474` says `disable_cuda_graph=False` enables a captured depth step, but capture isn't implemented (`modeling.py:9`, `model.py:344` are TODOs) — `true` is the only valid path today.
4. **Tests + docs.** `model.py` / `stages.py` (the per-row temp/top_k + row-pool batching glue) have no unit test, though Higgs ships `test_batched_step.py` for exactly that surface; the `generate_frame` math itself is well-pinned (greedy bit-parity vs HF, static-KV-vs-dense). Also: PR body says "3-stage" but it's 4 (`preprocessing → audio_encoder → tts_engine → vocoder`); the "follows Higgs file-for-file" claim breaks in a few places (strict-`raise` weight loader vs passthrough, `top_k` `<=` vs `<`, no cookbook page / H100 sizing) — softening it resolves several small notes.

Mergeable with (1) and (2) at minimum. Numbers reproducible on the `eustlb` mirror; happy to re-run on `sesame/csm-1b` once access is sorted.

**Comment by Venkat2811** [comment]
thx for the review @GaokaiZhang yes i didn't test againt dc gpus bcos of availability, will address n update PR

**Comment by chuyi-2025** [comment]
Hi@Venkat2811 — thanks for the CSM-1B work here.

I have access to an H200 and would like to help land the remaining review blockers from @GaokaiZhang’s H100 eval:

1. Hopper cuDNN SDPA crash in the depth-decoder attention path
2. Plain-TTS non-termination / high WER (~1/6 requests hitting the frame cap)

Before I start coding: would you prefer that I 
(a) collaborate on this PR (you grant push access / I open a PR into your branch), or 
(b) open a separate follow-up PR on top of `#800` once the fixes are ready?

@GaokaiZhang — to confirm acceptance criteria for a re-run: SeedTTS EN full set (1088), `--no-ref-audio`, `--max-concurrency 16`, WER via `openai/whisper-large-v3`, same as your earlier report? I’ll report `1×H200` numbers with the same harness.

Happy to coordinate either way. Thanks!

**Comment by Venkat2811** [comment]
@chuyi-2025 Thank you for checking & interest to collaborate. I got busy with other things & unfortunately can't work on this for quite some time. Please feel free to take the work done in this PR / start from scratch.

**Comment by Venkat2811** [comment]
@chuyi-2025 option b is simpler for both of us. pls feel free to take useful changes from this pr if any. in your new pr just refer to this one

**Comment by chuyi-2025** [comment]
Thanks @Venkat2811  — I’ll go with option (b) and open a new follow-up PR based on useful parts of #800.
I’ll ping you here once it’s up. Appreciate the handoff!
